File: app/src/main/java/com/gmail/fattazzo/formula1world/fragments/stats/season/SeasonComparisonStatsFragment.java
package com.gmail.fattazzo.formula1world.fragments.stats.season;

import android.support.v4.app.Fragment;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.gmail.fattazzo.formula1world.R;
import com.gmail.fattazzo.formula1world.ergast.imagedb.service.stats.StatsSeasonComparisonData;
import com.gmail.fattazzo.formula1world.ergast.imagedb.service.stats.StatsSeasonData;
import com.gmail.fattazzo.formula1world.service.StatisticsService;

import org.androidannotations.annotations.AfterViews;
import org.androidannotations.annotations.Bean;
import org.androidannotations.annotations.EFragment;
import org.androidannotations.annotations.FragmentArg;
import org.androidannotations.annotations.ViewById;

/**
 * @author fattazzo
 *         <p/>
 *         date: 07/09/17
 */
@EFragment(R.layout.fragment_stats_season_comparison)
public class SeasonComparisonStatsFragment extends Fragment {

    @Bean
    StatisticsService statisticsService;

    @FragmentArg
    int season;

    @ViewById
    TextView prevSeasonTV, seasonTV, nextSeasonTV;

    @ViewById
    ProgressBar prevRacesCompletedBar, racesCompletedBar, nextRacesCompletedBar;

    @ViewById
    LinearLayout seasonComparisonStatsContainer;

    @AfterViews
    void init() {
        StatsSeasonComparisonData stats = statisticsService.loadSeasonComparison(season);

        // Season ------------------------------------------------------
        prevSeasonTV.setText(String.valueOf(stats.getPrevSeasonData().getSeason()));
        seasonTV.setText(String.valueOf(stats.getSeasonData().getSeason()));
        nextSeasonTV.setText(String.valueOf(stats.getNextSeasonData().getSeason()));


        // Races completed ---------------------------------------------
        prevRacesCompletedBar.setMax(stats.getPrevSeasonData().getRounds());
        prevRacesCompletedBar.setProgress(stats.getPrevSeasonData().getRacesCompleted());
        racesCompletedBar.setMax(stats.getSeasonData().getRounds());
        racesCompletedBar.setProgress(stats.getSeasonData().getRacesCompleted());
        nextRacesCompletedBar.setMax(stats.getNextSeasonData().getRounds());
        nextRacesCompletedBar.setProgress(stats.getNextSeasonData().getRacesCompleted());

        seasonComparisonStatsContainer.removeAllViews();

        StatsSeasonData prevSeason = stats.getPrevSeasonData();
        StatsSeasonData season = stats.getSeasonData();
        StatsSeasonData nextSeason = stats.getNextSeasonData();
        // Rounds ------------------------------------------------------
        SeasonComparisonRowView roundsRow = SeasonComparisonRowView_.build(getContext(), R.string.rounds, prevSeason.getRounds(), season.getRounds(), nextSeason.getRounds());
        seasonComparisonStatsContainer.addView(roundsRow, new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));

        // Drivers -----------------------------------------------------
        SeasonComparisonRowView driversRow = SeasonComparisonRowView_.build(getContext(), R.string.drivers, prevSeason.getDrivers(), season.getDrivers(), nextSeason.getDrivers());
        seasonComparisonStatsContainer.addView(driversRow, new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));

        // Constructors -----------------------------------------------------
        SeasonComparisonRowView constructorsRow = SeasonComparisonRowView_.build(getContext(), R.string.constructors, prevSeason.getConstructors(), season.getConstructors(), nextSeason.getConstructors());
        seasonComparisonStatsContainer.addView(constructorsRow, new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));

        // WinningDrivers -----------------------------------------------------
        SeasonComparisonRowView winningDriversRow = SeasonComparisonRowView_.build(getContext(), R.string.winningDrivers, prevSeason.getWinningDrivers(), season.getWinningDrivers(), nextSeason.getWinningDrivers());
        seasonComparisonStatsContainer.addView(winningDriversRow, new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));

        // WinningConstructors -----------------------------------------------------
        SeasonComparisonRowView winningConstructorsRow = SeasonComparisonRowView_.build(getContext(), R.string.winningConstructors, prevSeason.getWinningConstructos(), season.getWinningConstructos(), nextSeason.getWinningConstructos());
        seasonComparisonStatsContainer.addView(winningConstructorsRow, new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));

        // PodiumDrivers -----------------------------------------------------
        SeasonComparisonRowView podiumDriversRow = SeasonComparisonRowView_.build(getContext(), R.string.podiumDrivers, prevSeason.getPodiumDrivers(), season.getPodiumDrivers(), nextSeason.getPodiumDrivers());
        seasonComparisonStatsContainer.addView(podiumDriversRow, new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));

        // PodiumConstructors -----------------------------------------------------
        SeasonComparisonRowView podiumConstructorsRow = SeasonComparisonRowView_.build(getContext(), R.string.podiumConstructors, prevSeason.getPodiumConstructors(), season.getPodiumConstructors(), nextSeason.getPodiumConstructors());
        seasonComparisonStatsContainer.addView(podiumConstructorsRow, new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));

        // WinningDriversPoints -----------------------------------------------------
        SeasonComparisonRowView winningDriversPointsRow = SeasonComparisonRowView_.build(getContext(), R.string.winningDriverPoints, prevSeason.getWinningDriverPoints(), season.getWinningDriverPoints(), nextSeason.getWinningDriverPoints());
        seasonComparisonStatsContainer.addView(winningDriversPointsRow, new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));

        // PointsAssigned -----------------------------------------------------
        SeasonComparisonRowView pointsAssignedRow = SeasonComparisonRowView_.build(getContext(), R.string.pointsAssigned, prevSeason.getPointsAssigned(), season.getPointsAssigned(), nextSeason.getPointsAssigned());
        seasonComparisonStatsContainer.addView(pointsAssignedRow, new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));
    }
}

File: app/src/main/kotlin/com/gmail/fattazzo/formula1world/fragments/stats/season/SeasonComparisonStatsFragment.kt
/*
 * Project: total-gp-world
 * File: SeasonComparisonStatsFragment.kt
 *
 * Created by fattazzo
 * Copyright © 2018 Gianluca Fattarsi. All rights reserved.
 *
 * MIT License
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.gmail.fattazzo.formula1world.fragments.stats.season

import android.support.v4.app.Fragment
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.TextView
import com.gmail.fattazzo.formula1world.R
import com.gmail.fattazzo.formula1world.service.StatisticsService
import org.androidannotations.annotations.*

/**
 * @author fattazzo
 *
 *
 * date: 07/09/17
 */
@EFragment(R.layout.fragment_stats_season_comparison)
open class SeasonComparisonStatsFragment : Fragment() {

    @Bean
    lateinit internal var statisticsService: StatisticsService

    @FragmentArg
    @JvmField protected var season: Int? = 0

    @ViewById
    lateinit internal var prevSeasonTV: TextView
    @ViewById
    lateinit internal var seasonTV: TextView
    @ViewById
    lateinit internal var nextSeasonTV: TextView

    @ViewById
    lateinit internal var prevRacesCompletedBar: ProgressBar
    @ViewById
    lateinit internal var racesCompletedBar: ProgressBar
    @ViewById
    lateinit internal var nextRacesCompletedBar: ProgressBar

    @ViewById
    lateinit internal var seasonComparisonStatsContainer: LinearLayout

    @AfterViews
    internal fun init() {
        val stats = statisticsService!!.loadSeasonComparison(this!!.season!!)

        // Season ------------------------------------------------------
        prevSeasonTV!!.text = stats.prevSeasonData.season.toString()
        seasonTV!!.text = stats.seasonData.season.toString()
        nextSeasonTV!!.text = stats.nextSeasonData.season.toString()


        // Races completed ---------------------------------------------
        prevRacesCompletedBar!!.max = stats.prevSeasonData.rounds
        prevRacesCompletedBar!!.progress = stats.prevSeasonData.racesCompleted
        racesCompletedBar!!.max = stats.seasonData.rounds
        racesCompletedBar!!.progress = stats.seasonData.racesCompleted
        nextRacesCompletedBar!!.max = stats.nextSeasonData.rounds
        nextRacesCompletedBar!!.progress = stats.nextSeasonData.racesCompleted

        seasonComparisonStatsContainer!!.removeAllViews()

        val prevSeason = stats.prevSeasonData
        val season = stats.seasonData
        val nextSeason = stats.nextSeasonData
        // Rounds ------------------------------------------------------
        val roundsRow = SeasonComparisonRowView_.build(context, R.string.rounds, prevSeason.rounds, season.rounds, nextSeason.rounds)
        seasonComparisonStatsContainer!!.addView(roundsRow, LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT))

        // Drivers -----------------------------------------------------
        val driversRow = SeasonComparisonRowView_.build(context, R.string.drivers, prevSeason.drivers, season.drivers, nextSeason.drivers)
        seasonComparisonStatsContainer!!.addView(driversRow, LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT))

        // Constructors -----------------------------------------------------
        val constructorsRow = SeasonComparisonRowView_.build(context, R.string.constructors, prevSeason.constructors, season.constructors, nextSeason.constructors)
        seasonComparisonStatsContainer!!.addView(constructorsRow, LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT))

        // WinningDrivers -----------------------------------------------------
        val winningDriversRow = SeasonComparisonRowView_.build(context, R.string.winningDrivers, prevSeason.winningDrivers, season.winningDrivers, nextSeason.winningDrivers)
        seasonComparisonStatsContainer!!.addView(winningDriversRow, LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT))

        // WinningConstructors -----------------------------------------------------
        val winningConstructorsRow = SeasonComparisonRowView_.build(context, R.string.winningConstructors, prevSeason.winningConstructos, season.winningConstructos, nextSeason.winningConstructos)
        seasonComparisonStatsContainer!!.addView(winningConstructorsRow, LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT))

        // PodiumDrivers -----------------------------------------------------
        val podiumDriversRow = SeasonComparisonRowView_.build(context, R.string.podiumDrivers, prevSeason.podiumDrivers, season.podiumDrivers, nextSeason.podiumDrivers)
        seasonComparisonStatsContainer!!.addView(podiumDriversRow, LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT))

        // PodiumConstructors -----------------------------------------------------
        val podiumConstructorsRow = SeasonComparisonRowView_.build(context, R.string.podiumConstructors, prevSeason.podiumConstructors, season.podiumConstructors, nextSeason.podiumConstructors)
        seasonComparisonStatsContainer!!.addView(podiumConstructorsRow, LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT))

        // WinningDriversPoints -----------------------------------------------------
        val winningDriversPointsRow = SeasonComparisonRowView_.build(context, R.string.winningDriverPoints, prevSeason.winningDriverPoints, season.winningDriverPoints, nextSeason.winningDriverPoints)
        seasonComparisonStatsContainer!!.addView(winningDriversPointsRow, LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT))

        // PointsAssigned -----------------------------------------------------
        val pointsAssignedRow = SeasonComparisonRowView_.build(context, R.string.pointsAssigned, prevSeason.pointsAssigned, season.pointsAssigned, nextSeason.pointsAssigned)
        seasonComparisonStatsContainer!!.addView(pointsAssignedRow, LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT))
    }
}
